control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go

```diff
@@ -1389,6 +1415,62 @@ func (r *reconciler) reconcileClusterVersion(ctx context.Context, hcp *hyperv1.H
 	return nil
 }
 
+func (r *reconciler) reconcileControlPlaneDataPlaneConnectivityConditions(ctx context.Context, hcp *hyperv1.HostedControlPlane, log logr.Logger) error {
+	condition := &metav1.Condition{
+		Type:    string(hyperv1.ControlPlaneToDataPlaneConnectivityHealthy),
+		Status:  metav1.ConditionUnknown,
+		Reason:  hyperv1.ControlPlaneToDataPlaneReasonNoWorkerNodesAvailable,
+		Message: "No worker nodes available",
+	}
+	totalNodes, err := util.CountAvailableNodes(ctx, r.client)
+	if err != nil {
+		log.Error(err, "unable to count worker nodes")
+	}
+	if totalNodes > 0 {
+		condition.Status = metav1.ConditionFalse
+		condition.Reason = hyperv1.ControlPlaneToDataPlaneReasonNoKonnectivityAgentPodsFound
+		condition.Message = "Couldn't find an konnectivity-agent running in data plane"
+		var podList corev1.PodList
+		if err := r.client.List(ctx, &podList,
+			client.MatchingLabels{"app": "konnectivity-agent"}, client.InNamespace("kube-system")); err != nil {
+			log.Error(err, "unable to list konnectivity-agent PODs in kube-system namespace")
+		}
+		for _, pod := range podList.Items {
```

Comment:
Thinking about it a little more. Should we verify that "all" konnectivity-agent Pods are running? It's running a DeamonSet and it's required on each node. The connectivity is not fully established until the konnectivity-pod is running on each Node. Say, we successfully start a konnectivity-agent Pod on one node but the other ones keep failing. Then we'll get success and the message "At least a konnectivity-agent is running on data plane and logs are accessible" even though konnectivity on other nodes is not yet available (and it might not be ever).

Reply:
@mgencur one konnectivity-agent POD running grant network connectivty between control-plane and data-plane